kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSet.kt

```diff
@@ -1,87 +1,137 @@
+                arrayOf(false, { m: Collection<*> -> as_Set1(m) }),
+                arrayOf(true, { m: Collection<*> -> mk_SetExtension(*m.toTypedArray()) }),
+                arrayOf(false, { m: Collection<*> -> mk_Set1Extension(*m.toTypedArray()) }),
+            )
+    }
+
+    private fun empty() = create<Any>()
+    private fun <T> create(vararg m: T) = creator.invoke(m.toList())
 
     @Test
     fun inter() {
-        assertEquals(mk_Set(), mk_Set<Int>() inter mk_Set())
-        assertEquals(mk_Set(), mk_Set(1, 2, 3) inter mk_Set())
-        assertEquals(mk_Set(2), mk_Set(1, 2, 3) inter mk_Set(2))
-        assertEquals(mk_Set(2), mk_Set(1, 2, 3) inter mk_Set(2, 4))
-        assertEquals(mk_Set(1, 2, 3), mk_Set(1, 2, 3) inter mk_Set(2, 4, 3, 1))
+        if (allowsEmpty) {
+            assertEquals(empty(), empty() inter empty())
+            assertEquals(empty(), create(1, 2, 3) inter empty())
+            assertEquals(empty(), create(1) inter create(2))
+        } else {
+            // inter creates a set of the same type as the first input
+            causesPreconditionFailure { create(1) inter create(2) }
+        }
+        assertEquals(create(2), create(1, 2, 3) inter create(2))
+        assertEquals(create(2), create(1, 2, 3) inter create(2, 4))
+        assertEquals(create(1, 2, 3), create(1, 2, 3) inter create(2, 4, 3, 1))
     }
 
     @Test
     fun union() {
-        assertEquals(mk_Set(), mk_Set<Int>() union mk_Set())
-        assertEquals(mk_Set(1, 2, 3), mk_Set(1, 2, 3) union mk_Set())
-        assertEquals(mk_Set(1, 2, 3), mk_Set(1, 2, 3) union mk_Set(2))
-        assertEquals(mk_Set(1, 2, 3, 4), mk_Set(1, 2, 3) union mk_Set(2, 4))
-        assertEquals(mk_Set(1, 2, 3, 4), mk_Set(1, 2, 3) union mk_Set(2, 4, 3, 1))
+        if (allowsEmpty) {
+            assertEquals(empty(), empty() union empty())
+        }
+        assertEquals(create(1, 2, 3), create(1, 2, 3) union create(2))
+        assertEquals(create(1, 2, 3, 4), create(1, 2, 3) union create(2, 4))
+        assertEquals(create(1, 2, 3, 4), create(1, 2, 3) union create(2, 4, 3, 1))
+        assertEquals(create(1, 2, 3), create(1, 2, 3) union mk_Set())
+
     }
 
     @Test
     fun subset() {
-        assertEquals(true, mk_Set<Nothing>() subset mk_Set())
-        assertEquals(true, mk_Set<Int>() subset mk_Set(1, 2, 3))
-        assertEquals(false, mk_Set<Int>(1) subset mk_Set())
-        assertEquals(true, mk_Set(1) subset mk_Set(1, 2, 3))
-        assertEquals(false, mk_Set(1, 2, 3) subset mk_Set(1))
-        assertEquals(false, mk_Set(2) subset mk_Set(mk_Set(1, 2, 3), mk_Set(2), mk_Set()))
-        assertEquals(true, mk_Set(mk_Set(3)) subset mk_Set(mk_Set(1, 2, 3), mk_Set(3), mk_Set()))
+        if (allowsEmpty) {
+            assertEquals(true, empty() subset empty())
+            assertEquals(false, create(2) subset create(create(1, 2, 3), create(2), empty()))
+            assertEquals(true, create(create(3)) subset create(create(1, 2, 3), create(3), empty()))
```

Comment:
⛏️ I don't think these two are dependent on the type of the empty set matching the type under test. I would just use `mk_Set()` instead of `empty()` and take those out of the conditional.

Reply:
:no_entry_sign: Indeed the type of the set (incl set vs set1) is only important for equality if you've explicitly asked for it be, so this is a bit misleading and I'd prefer it removed

Reply:
I guess I wasn't clear :smile: .. I meant remove the use of `empty()` (I actually meant remove the use of empty everywhere!) not the tests